core/modules/xrdsvc/StreamBuffer.h

```diff
@@ -0,0 +1,96 @@
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_XRDSVC_STREAMBUFFER_H
+#define LSST_QSERV_XRDSVC_STREAMBUFFER_H
+
+// System headers
+#include <atomic>
+#include <condition_variable>
+#include <deque>
+#include <mutex>
+#include <string>
+
+// qserv headers
+#include "util/InstanceCount.h"
+
+// Third-party headers
+#include "XrdSsi/XrdSsiErrInfo.hh" // required by XrdSsiStream
+#include "XrdSsi/XrdSsiStream.hh"
+
+namespace lsst {
+namespace qserv {
+namespace xrdsvc {
+
+
+/// StreamBuffer is a single use buffer for transferring data packets
+/// to XrdSsi.
+/// Its notable feature is the Recycle() function, which XrdSsi will
+/// promptly call when it no longer needs the buffer.
+class StreamBuffer : public XrdSsiStream::Buffer {
+public:
+    using Ptr = std::shared_ptr<StreamBuffer>;
+
+    // Copying this would be very confusing for something waiting for Recycle().
+    StreamBuffer() = delete;
+    StreamBuffer(StreamBuffer const&) = delete;
+    StreamBuffer& operator=(StreamBuffer const&) = delete;
+
+    /// Factory function, because this should be able to delete itself when Recycle() is called.
+    //  The constructor uses move to avoid copying the string.
+    static StreamBuffer::Ptr createWithMove(std::string &input);
+
+    size_t getSize() const { return _dataStr.size(); }
+
+    /// @Return total number of bytes used by ALL StreamBuffer objects.
+    static size_t getTotalBytes() { return _totalBytes; }
+
+    //!> Call to recycle the buffer when finished
+    void Recycle() override;
+
+    // Wait until recycle is called.
+    void waitForDoneWithThis();
+
+    // Inherited from XrdSsiStream:
+    // char  *data; //!> -> Buffer containing the data
+    // Buffer *next; //!> For chaining by buffer receiver
+
+    virtual ~StreamBuffer();
+
+private:
+    // This constructor will invalidate 'input'.
+    explicit StreamBuffer(std::string &input);
+
+    std::string _dataStr;
+    std::mutex _mtx;
+    std::condition_variable _cv;
+    bool doneWithThis{false};
+    Ptr _selfKeepAlive; ///< keep this object alive until after Recycle() is called.
+    util::InstanceCount _ic{"StreamBuffer"}; ///< Useful as it indicates amount of waiting for czar.
```

Comment:
Never seen `InstanceCount` class before. If it does what I think it does it would be pretty inefficient.

Reply:
Agreed. The only reason why one may want to maintain such registry of instances would be if someone wanted to see counters for all of them. Otherwise a template version would work just great (and very fast):
template < class T >
struct InstanceCounter {
InstanceCounter() { count++; }
static std::atomic<unsigned int> count{0};
};

Then replace the member variable in class StreamBuffer with:

util::InstanceCount < StreamBuffer > _ic;

Reply:
Every time InstantanceCount's destructor is called, it prints a list of all the items in its map. This is very useful when you are looking at a log file. It's pretty easy to notice that, say StreamBuffer has not dropped to zero when you expected it to, which was the case a couple of times during testing. The number of Tasks dropped to zero, but there were a few thousand StreamBuffers, which was blindingly obvious in a single log message.  Otherwise, I would of really had to of gone looking for it. The other advantage of the string based version is you can stick it in functions and find out how many threads are in the function at the same time very easily.

I can probably combine the 2 approaches. The map is usually very small, less than 5 elements, and I'm not sure it's better to lock one mutex and search a small map, or print the values of 5 std::atomic's with the underlying locking that occurs.

Reply:
Last item first, access to simple atomic variables (i.e. native types) is very fast as the compiler uses native machine instructions to do this and no locking occurs.  You can write a little test program and use std::atomic var->is_lock_free(). or since C++11 atomic_is_lock_free( const std::atomic<T>* obj ). If something is not lock free then it all depends on how much contention there is on the variable. If it's likely to be locked by someone else then doing 5 vs one is very expensive. Otherwise, it's a toss up.

As for the first, I'm curious to know why buffers get stacked up especaly when the thread count drops to zero. Doesn't soud like something that should happen. Why does it?

Reply:
Something strange may happen when the czar cancels a user query because the result was too large and there are a couple of other select count(*) shared scans running. Somehow several buffers wind up in limbo. I have some ideas, but not much time. I'm going to look into in the next ticket.